File: lib/msf/core/modules/external/python/metasploit/module.py
import json
import logging
import os
import sys

from metasploit import cli

__CLI_MODE__ = False


class LogFormatter(logging.Formatter):
    def __init__(self, prefix, *args, **kwargs):
        super(LogFormatter, self).__init__(*args, **kwargs)
        self.prefix = prefix

    def format(self, record):
        return self.prefix + record.msg


class LogHandler(logging.Handler):
    def emit(self, record):
        level = 'debug'
        if record.levelno >= logging.ERROR:
            level = 'error'
        elif record.levelno >= logging.WARNING:
            level = 'warning'
        elif record.levelno >= logging.INFO:
            level = 'info'
        log(self.format(record), level)
        return

    @classmethod
    def setup(cls, level=logging.DEBUG, name=None, msg_prefix=None):
        logger = logging.getLogger(name)
        handler = cls()

        if level is not None:
            logger.setLevel(level)
        if msg_prefix is not None:
            handler.setFormatter(LogFormatter(msg_prefix))
        logger.addHandler(handler)
        return handler

def log(message, level='info'):
    if not __CLI_MODE__:
        rpc_send({'jsonrpc': '2.0', 'method': 'message', 'params': {
            'level': level,
            'message': message
        }})
    else:
        cli.log(message, level)


def report_host(ip, **opts):
    host = opts.copy()
    host.update({'host': ip})
    report('host', host)


def report_service(ip, **opts):
    service = opts.copy()
    service.update({'host': ip})
    report('service', service)


def report_vuln(ip, name, **opts):
    vuln = opts.copy()
    vuln.update({'host': ip, 'name': name})
    report('vuln', vuln)


def report_correct_password(username, password, **opts):
    info = opts.copy()
    info.update({'username': username, 'password': password})
    report('correct_password', info)


def report_wrong_password(username, password, **opts):
    info = opts.copy()
    info.update({'username': username, 'password': password})
    report('wrong_password', info)


def run(metadata, module_callback, soft_check=None):
    global __CLI_MODE__

    caps = []
    if soft_check:
        caps.append('soft_check')

    meta = metadata.copy()
    meta.update({'capabilities': caps})

    if len(sys.argv) > 1:
        __CLI_MODE__ = True

    req = None
    if __CLI_MODE__:
        req = cli.parse(meta)
    else:
        req = json.loads(os.read(0, 10000).decode("utf-8"))

    callback = None
    if req['method'] == 'describe':
        rpc_send({'jsonrpc': '2.0', 'id': req['id'], 'result': meta})
    elif req['method'] == 'soft_check':
        if soft_check:
            callback = soft_check
        else:
            rpc_send({'jsonrpc': '2.0', 'id': req['id'], 'error': {'code': -32601, 'message': 'Soft checks are not supported'}})
    elif req['method'] == 'run':
        callback = module_callback

    if callback:
        args = req['params']
        ret = callback(args)
        if ret and __CLI_MODE__:
            cli.ret(ret)

        rpc_send({'jsonrpc': '2.0', 'id': req['id'], 'result': {
            'message': 'Module completed',
            'return': ret
        }})


def report(kind, data):
    if not __CLI_MODE__:
        rpc_send({'jsonrpc': '2.0', 'method': 'report', 'params': {
            'type': kind, 'data': data
        }})
    else:
        cli.report(kind, data)


def rpc_send(req):
    # Silently ignore when run manually, the calling code should know how to
    # handle if it is important
    if not __CLI_MODE__:
        print(json.dumps(req))
        sys.stdout.flush()

File: lib/msf/core/modules/external/go/src/metasploit/module.go
package metasploit

import (
	"bufio"
	"encoding/json"
	"log"
	"os"
	"strings"
)

type response struct {
	Jsonrpc string `json:"jsonrpc"`
	ID      string `json:"id"`
}

func rpcSend(res interface{}) error {
	resStr, err := json.Marshal(res)
	if err != nil {
		return err
	}
	f := bufio.NewWriter(os.Stdout)
	if _, err := f.Write(resStr); err != nil {
		return err
	}
	if err := f.Flush(); err != nil {
		return err
	}

	return nil
}

type (
	logparams struct {
		Level   string `json:"level"`
		Message string `json:"message"`
	}

	logRequest struct {
		Jsonrpc string    `json:"jsonrpc"`
		Method  string    `json:"method"`
		Params  logparams `json:"params"`
	}
)

func Log(message string, level string) {
	req := &logRequest{"2.0", "message", logparams{level, message}}
	if err := rpcSend(req); err != nil {
		log.Fatal(err)
	}
}

type (
	reportparams struct {
		Type string            `json:"type"`
		Data map[string]string `json:"data"`
	}

	reportRequest struct {
		Jsonrpc string       `json:"jsonrpc"`
		Method  string       `json:"method"`
		Params  reportparams `json:"params"`
	}
)

func report(kind string, base map[string]string, opts map[string]string) error {
	for k, v := range base {
		opts[k] = v
	}
	req := &reportRequest{"2.0", "report", reportparams{kind, opts}}
	return rpcSend(req)
}

func ReportHost(ip string, opts map[string]string) {
	base := map[string]string{"host": ip}
	if err := report("host", base, opts); err != nil {
		log.Fatal(err)
	}
}

func ReportService(ip string, opts map[string]string) {
	base := map[string]string{"host": ip}
	if err := report("service", base, opts); err != nil {
		log.Fatal(err)
	}
}

func ReportVuln(ip string, name string, opts map[string]string) {
	base := map[string]string{"host": ip, "name": name}
	if err := report("vuln", base, opts); err != nil {
		log.Fatal(err)
	}
}

func ReportCorrectPassword(username string, password string, opts map[string]string) {
	base := map[string]string{"username": username, "password": password}
	if err := report("correct_password", base, opts); err != nil {
		log.Fatal(err)
	}
}

func ReportWrongPassword(username string, password string, opts map[string]string) {
	base := map[string]string{"username": username, "password": password}
	if err := report("wrong_password", base, opts); err != nil {
		log.Fatal(err)
	}
}

type (
	Reference struct {
		Type string `json:"type"`
		Ref  string `json:"ref"`
	}

	Target struct {
		Platform string `json:"platform"`
		Arch     string `json:"arch"`
	}

	Option struct {
		Type        string `json:"type"`
		Description string `json:"description"`
		Required    bool   `json:"required"`
		Default     string `json:"default"`
	}

	Metadata struct {
		Name         string              `json:"name"`
		Description  string              `json:"description"`
		Authors      []string            `json:"authors"`
		Date         string              `json:"date"`
		References   []Reference         `json:"references"`
		Type         string              `json:"type"`
		Rank         string              `json:"rank"`
		WFSDelay     int                 `json:"wfsdelay"`
		Privileged   bool                `json:"privileged"`
		Targets      []Target            `json:"targets,omitempty"`
		Capabilities []string            `json:"capabilities"`
		Payload      map[string]string   `json:"payload,omitempty"`
		Options      map[string]Option   `json:"options,omitempty"`
		Notes        map[string][]string `json:"notes,omitempty"`
	}

	Request struct {
		Jsonrpc string `json:"jsonrpc"`
		Method  string `json:"method"`
		ID      string `json:"id"`
	}

	MetadataResponse struct {
		Jsonrpc string    `json:"jsonrpc"`
		ID      string    `json:"id"`
		Result  *Metadata `json:"result"`
	}

	RunResult struct {
		Message string `json:"message"`
		Return  string `json:"return"`
	}

	RunResponse struct {
		Jsonrpc string    `json:"jsonrpc"`
		ID      string    `json:"id"`
		Result  RunResult `json:"result"`
	}
)

// RunCallback represents the exploit method to call from the module
type RunCallback func(req *Request) string

// Run runs the exploit
func Run(metadata *Metadata, callback RunCallback) {
	var req Request

	err := json.NewDecoder(os.Stdin).Decode(&req)
	if err != nil {
		log.Fatalf("could not decode JSON: %v", err)
	}

	switch strings.ToLower(req.Method) {
	case "describe":
		metadata.Capabilities = []string{"run"}
		res := &MetadataResponse{"2.0", req.ID, metadata}
		if err := rpcSend(res); err != nil {
			log.Fatalf("error on running %s: %v", req.Method, err)
		}
	case "run":
		ret := callback(&req)
		res := &RunResponse{"2.0", req.ID, RunResult{"Module complete", ret}}
		if err := rpcSend(res); err != nil {
			log.Fatalf("error on running %s: %v", req.Method, err)
		}
	default:
		log.Fatalf("method %s not implemented yet", req.Method)
	}
}
